File: cs/src/Management/TunnelAccessTokenProperties.cs
// <copyright file="TunnelAccessTokenProperties.cs" company="Microsoft">
// Copyright (c) Microsoft. All rights reserved.
// </copyright>

using System;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Microsoft.DevTunnels.Management;

/// <summary>
/// Supports parsing tunnel access token JWT properties to allow for some pre-validation
/// and diagnostics.
/// </summary>
/// <remarks>
/// Applications generally should not attempt to interpret or rely on any token properties
/// other than <see cref="TunnelAccessTokenProperties.Expiration" />, because the service
/// may change or omit those claims in the future. Other claims are exposed here only for
/// diagnostic purposes.
/// </remarks>
public class TunnelAccessTokenProperties
{
    private const string ClusterIdClaimName = "clusterId";
    private const string TunnelIdClaimName = "tunnelId";
    private const string TunnelPortClaimName = "tunnelPort";
    private const string ScopeClaimName = "scp";
    private const string IssuerClaimName = "iss";
    private const string ExpirationClaimName = "exp";

    /// <summary>
    /// Gets the token cluster ID claim.
    /// </summary>
    public string? ClusterId { get; private set; }

    /// <summary>
    /// Gets the token tunnel ID claim.
    /// </summary>
    public string? TunnelId { get; private set; }

    /// <summary>
    /// Gets the token tunnel ports claim.
    /// </summary>
    public ushort[]? TunnelPorts { get; private set; }

    /// <summary>
    /// Gets the token scopes claim.
    /// </summary>
    public string[]? Scopes { get; private set; }

    /// <summary>
    /// Gets the token issuer URI.
    /// </summary>
    public string? Issuer { get; private set; }

    /// <summary>
    /// Gets the token expiration.
    /// </summary>
    [JsonIgnore]
    public DateTime? Expiration { get; private set; }

    /// <inheritdoc/>
    public override string ToString()
    {
        var s = new StringBuilder();

        if (!string.IsNullOrEmpty(TunnelId))
        {
            s.Append("tunnel=");
            s.Append(TunnelId);

            if (!string.IsNullOrEmpty(ClusterId))
            {
                s.Append('.');
                s.Append(ClusterId);
            }
        }

        if (TunnelPorts != null)
        {
            if (s.Length > 0) s.Append(", ");
            if (TunnelPorts.Length == 1)
            {
                s.Append("port=");
                s.Append(TunnelPorts[0]);
            }
            else
            {
                s.AppendFormat("ports=[{0}]", string.Join(", ", TunnelPorts));
            }
        }

        var scopes = Scopes;
        if (scopes != null)
        {
            if (s.Length > 0) s.Append(", ");
            s.AppendFormat("scopes=[{0}]", string.Join(", ", scopes));
        }

        if (!string.IsNullOrEmpty(Issuer))
        {
            if (s.Length > 0) s.Append(", ");
            s.Append("issuer=");
            s.Append(Issuer);
        }

        var expiration = Expiration;
        if (expiration != null)
        {
            if (s.Length > 0) s.Append(", ");

            // Get the current date-time without fractional seconds.
            var nowTicks = DateTime.UtcNow.Ticks;
            var now = new DateTime(nowTicks - nowTicks % 10000000, DateTimeKind.Utc);

            var lifetime = expiration.Value >= now
                ? (expiration.Value - now).ToString() + " remaining"
                : (now - expiration.Value) + " ago";
            s.AppendFormat("expiration={0:s}Z ({1})", expiration.Value, lifetime);
        }

        return s.ToString();
    }

    /// <summary>
    /// Checks if the tunnel access token expiration claim is in the past.
    /// </summary>
    /// <exception cref="UnauthorizedAccessException">The token is expired.</exception>
    /// <remarks>
    /// Note this does not throw if the token is an invalid format.
    /// Uses the client's system time for validation.
    /// </remarks>
    public static void ValidateTokenExpiration(string token)
    {
        var t = TryParse(token);
        if (t?.Expiration <= DateTime.UtcNow)
        {
            throw new UnauthorizedAccessException("The access token is expired: " + t);
        }
    }

    /// <summary>
    /// Gets token representation for tracing.
    /// </summary>
    public static string GetTokenTrace(string? token)
    {
        if (token == null)
        {
            return "<null>";
        }

        if (token == string.Empty)
        {
            return "<empty>";
        }

        return TryParse(token) is TunnelAccessTokenProperties t ? $"<JWT: {t}>" : "<token>";
    }

    /// <summary>
    /// Attempts to parse a tunnel access token (JWT). This does NOT validate the token
    /// signature or any claims.
    /// </summary>
    /// <returns>The parsed token properties, or null if the token is an invalid format.</returns>
    /// <remarks>
    /// Applications generally should not attempt to interpret or rely on any token properties
    /// other than <see cref="Expiration" />, because the service may change or omit those claims
    /// in the future. Other claims are exposed here only for diagnostic purposes.
    /// </remarks>
    public static TunnelAccessTokenProperties? TryParse(string token)
    {
        Requires.NotNullOrEmpty(token, nameof(token));

        // JWTs are encoded in 3 parts: header, body, and signature.
        var tokenParts = token.Split('.');
        if (tokenParts.Length != 3)
        {
            return null;
        }

        var tokenBodyJson = Base64UrlDecode(tokenParts[1]);
        if (tokenBodyJson == null)
        {
            return null;
        }

        try
        {
            var tokenElement = JsonSerializer.Deserialize<JsonElement>(tokenBodyJson)!;
            var tokenProperties = new TunnelAccessTokenProperties();

            if (tokenElement.TryGetProperty(ClusterIdClaimName, out var clusterIdElement))
            {
                tokenProperties.ClusterId = clusterIdElement.GetString();
            }

            if (tokenElement.TryGetProperty(TunnelIdClaimName, out var tunnelIdElement))
            {
                tokenProperties.TunnelId = tunnelIdElement.GetString();
            }

            if (tokenElement.TryGetProperty(TunnelPortClaimName, out var tunnelPortElement))
            {
                // The port claim value may be a single port number or an array of ports.
                if (tunnelPortElement.ValueKind == JsonValueKind.Array)
                {
                    var array = new ushort[tunnelPortElement.GetArrayLength()];

                    for (int i = 0; i < array.Length; i++)
                    {
                        array[i] = tunnelPortElement[i].GetUInt16();
                    }

                    tokenProperties.TunnelPorts = array;
                }
                else if (tunnelPortElement.ValueKind == JsonValueKind.Number)
                {
                    tokenProperties.TunnelPorts = new[] { tunnelPortElement.GetUInt16() };
                }
            }

            if (tokenElement.TryGetProperty(ScopeClaimName, out var scopeElement))
            {
                var scopes = scopeElement.GetString();
                tokenProperties.Scopes = string.IsNullOrEmpty(scopes) ? null : scopes.Split(' ');
            }

            if (tokenElement.TryGetProperty(IssuerClaimName, out var issuerElement))
            {
                tokenProperties.Issuer = issuerElement.GetString();
            }

            if (tokenElement.TryGetProperty(ExpirationClaimName, out var expirationElement) &&
                expirationElement.ValueKind == JsonValueKind.Number)
            {
                var exp = expirationElement.GetInt64();
                tokenProperties.Expiration = DateTimeOffset.FromUnixTimeSeconds(exp).UtcDateTime;
            }

            return tokenProperties;
        }
        catch (JsonException)
        {
            return null;
        }
    }

    private static string? Base64UrlDecode(string encodedString)
    {
        // Convert from base64url encoding to base64 encoding: replace chars and add padding.
        encodedString = encodedString.Replace('-', '+').Replace('_', '/');
        encodedString += new string('=', 3 - ((encodedString.Length - 1) % 4));

        try
        {
            var bytes = Convert.FromBase64String(encodedString);
            var result = Encoding.UTF8.GetString(bytes);
            return result;
        }
        catch (FormatException)
        {
            return null;
        }
    }
}

File: ts/src/management/tunnelAccessTokenProperties.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Tunnel } from '@microsoft/dev-tunnels-contracts';

/**
 * Supports parsing tunnel access token JWT properties to allow for some pre-validation
 * and diagnostics.
 *
 * Applications generally should not attempt to interpret or rely on any token properties
 * other than `expiration`, because the service may change or omit those claims in the future.
 * Other claims are exposed here only for diagnostic purposes.
 */
export class TunnelAccessTokenProperties {
    private constructor(
        public readonly clusterId?: string,
        public readonly tunnelId?: string,
        public readonly tunnelPorts?: number[],
        public readonly scopes?: string[],
        public readonly issuer?: string,
        public readonly expiration?: Date,
    ) {}

    public toString(): string {
        let s = '';

        if (this.tunnelId) {
            s += 'tunnel=';
            s += this.tunnelId;

            if (this.clusterId) {
                s += '.';
                s += this.clusterId;
            }
        }

        if (this.tunnelPorts && this.tunnelPorts.length > 0) {
            if (s.length > 0) s += ', ';
            if (this.tunnelPorts.length === 1) {
                s += `port=${this.tunnelPorts[0]}`;
            } else {
                s += `ports=[${this.tunnelPorts.join(', ')}]`;
            }
        }

        if (this.scopes) {
            if (s.length > 0) s += ', ';
            s += `scopes=[${this.scopes.join(', ')}]`;
        }

        if (this.issuer) {
            if (s.length > 0) s += ', ';
            s += 'issuer=';
            s += this.issuer;
        }

        if (this.expiration) {
            if (s.length > 0) s += ', ';
            s += `expiration=${this.expiration.toString().replace('.000Z', 'Z')}`;
        }

        return s;
    }

    /**
     * Checks if the tunnel access token expiration claim is in the past.
     *
     * (Does not throw if the token is an invalid format.)
     */
    public static validateTokenExpiration(token: string): void {
        const t = TunnelAccessTokenProperties.tryParse(token);
        if (t?.expiration) {
            if (t.expiration < new Date()) {
                throw new Error('The access token is expired: ' + t);
            }
        }
    }

    /**
     * Attempts to parse a tunnel access token (JWT). This does NOT validate the token
     * signature or any claims.
     */
    public static tryParse(token: string): TunnelAccessTokenProperties | null {
        if (typeof token !== 'string') throw new TypeError('Token string expected.');

        // JWTs are encoded in 3 parts: header, body, and signature.
        const tokenParts = token.split('.');
        if (tokenParts.length !== 3) {
            return null;
        }

        const tokenBodyJson = TunnelAccessTokenProperties.base64UrlDecode(tokenParts[1]);
        if (!tokenBodyJson) {
            return null;
        }

        try {
            const tokenJson = JSON.parse(tokenBodyJson);
            const clusterId: string | undefined = tokenJson.clusterId;
            const tunnelId: string | undefined = tokenJson.tunnelId;
            const ports: number | number[] | undefined = tokenJson.tunnelPorts;
            const scp: string | undefined = tokenJson.scp;
            const iss: string | undefined = tokenJson.iss;
            const exp: number | undefined = tokenJson.exp;

            return new TunnelAccessTokenProperties(
                clusterId,
                tunnelId,
                typeof ports === 'number' ? [ports] : ports,
                scp?.split(' '),
                iss,
                typeof exp === 'number' ? new Date(exp * 1000) : undefined,
            );
        } catch {
            return null;
        }
    }

    /**
     * Gets the tunnal access token trace string.
     * 'none' if null or undefined, parsed token info if can be parsed, or 'token' if cannot be parsed.
     */
    public static getTokenTrace(token?: string | null | undefined): string {
        return !token ? 'none' : TunnelAccessTokenProperties.tryParse(token)?.toString() ?? 'token';
    }

    /**
     * Gets a tunnel access token that matches any of the provided access token scopes.
     * Validates token expiration if the token is found and throws an error if it's expired.
     * @param tunnel The tunnel to get the access tokens from.
     * @param accessTokenScopes What scopes the token needs to have.
     * @returns Tunnel access token if found; otherwise, undefined.
     */
    public static getTunnelAccessToken(
        tunnel?: Tunnel | null,
        accessTokenScopes?: string | string[],
    ): string | undefined {
        if (!tunnel?.accessTokens || !accessTokenScopes) {
            return;
        }

        if (!Array.isArray(accessTokenScopes)) {
            accessTokenScopes = [accessTokenScopes];
        }

        for (const scope of accessTokenScopes) {
            for (const [key, accessToken] of Object.entries(tunnel.accessTokens)) {
                // Each key may be either a single scope or space-delimited list of scopes.
                if (accessToken && key.split(' ').includes(scope)) {
                    TunnelAccessTokenProperties.validateTokenExpiration(accessToken);
                    return accessToken;
                }
            }
        }
    }

    private static base64UrlDecode(encodedString: string): string | null {
        // Convert from base64url encoding to base64 encoding: replace chars and add padding.
        encodedString = encodedString.replace('-', '+');
        while (encodedString.length % 4 !== 0) {
            encodedString += '=';
        }

        try {
            const result = atob(encodedString);
            return result;
        } catch {
            return null;
        }
    }
}
